File: LeetCode/Hard/0068-text-justification/0068-text-justification.py
class Solution:
    def fullJustify(self, words: List[str], maxWidth: int) -> List[str]:
        ret = []
        left, right = 0, 0
        curr = 0
        n = len(words)
        
        while right < n:
            if len(words[right]) + curr <= maxWidth:
                curr += len(words[right]) + 1
                right += 1
            else:
                temp = ""
                if left + 1 == right:
                    temp += words[left] + " " * (maxWidth - curr + 1)
                else:
                    num_of_words = right - left
                    curr -= num_of_words
                    space_needed = maxWidth - curr
                    each_space = space_needed // (num_of_words - 1)
                    extra_space = space_needed % (num_of_words - 1)

                    for i in range(left, right - 1):
                        temp += words[i] + " " * each_space
                        if extra_space:
                            temp += " "
                            extra_space -= 1
                    temp += words[right-1]

                ret.append(temp)

                left = right
                curr = len(words[right]) + 1

                right += 1
        
        temp = words[left]

        for i in range(left + 1, right):
            temp += " " + words[i]

        temp += " " * (maxWidth - len(temp))

        ret.append(temp)

        return ret

File: LeetCode/Hard/0068-text-justification/0068-text-justification.cpp
class Solution {
public:
    vector<string> fullJustify(vector<string>& words, int maxWidth) {
        vector<string> ret;
        int l = 0, r = 0, curr = 0, n = words.size();

        string temp;

        while (r < n) {
            if (curr + words[r].length() <= maxWidth)
                curr += words[r++].length() + 1;
            else {
                if (l + 1 == r) {
                    temp.append(words[l]);
                    temp.append(string(maxWidth - curr + 1, ' '));
                } else {
                    int numOfWords = r - l;
                    int spaceNeeded = maxWidth - curr + numOfWords;
                    int spaceEach = spaceNeeded / (numOfWords - 1);
                    int extraSpace = spaceNeeded % (numOfWords - 1);

                    for (int i = l; i < r - 1; ++i) {
                        temp.append(words[i]);
                        temp.append(string(spaceEach, ' '));

                        if (extraSpace > 0) {
                            temp.append(" ");
                            extraSpace--;
                        }
                    }

                    temp.append(words[r - 1]);
                }

                ret.push_back(temp);
                temp.clear();
                l = r;
                curr = words[r++].length() + 1;
            }
        }

        temp.append(words[l]);

        for (int i = l + 1; i < r; ++i) {
            temp.append(" ");
            temp.append(words[i]);
        }

        temp.append(string(maxWidth - curr + 1, ' '));
        ret.push_back(temp);

        return ret;
    }
};